cpp/arcticdb/stream/merge.hpp

```diff
@@ -30,8 +44,33 @@ void do_merge(
             std::advance(val, IndexType::field_count());
             for(; val != next->row().end(); ++val) {
                 val->visit_field([&rb] (const auto& opt_v, std::string_view name, const TypeDescriptor& type_desc) {
-                    if(opt_v)
-                        rb.set_scalar_by_name(name, opt_v.value(), type_desc.data_type());
+                    if (opt_v) {
+                        const StreamDescriptor& descriptor = rb.descriptor();
+                        const std::optional<size_t> field_idx = descriptor.find_field(name);
+                        if (!field_idx || type_desc == descriptor.field(*field_idx).type()) {
```

Comment:
It would be better to check the whole column as we load the segment, since within each segment all the elements in a column will be the same type.

Reply:
This one is a bit tricky, since each staged segment can have different datatype and we're not processing segment by segment but row by row and consecutive rows can be from different segments. Anyway it's a valid concern, I'll think of some way to avoid it.

On a tangential line, can is possible to propagate if dynamic_schema is used to this function?